File: algorithms/quick_sort.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-  
#####################################################
#  
# File: quick_sort 
# Created: 2014-10-16 21:04:38
# Last Modified:  
#   
# Author: 杨秀隆 sndnyang <xiulong.yancey.yang@gmail.com>
#         sndnyang.github.io
# Description:
#  
#####################################################

def selection(data, l, h):
    pass

def partition_for(data, l, h):
    selection(data, l, h)
    pass

def partition(data, l, h):
    selection(data, l, h)

    guard = data[l]

    i, j = l+1, h
    while i<j:

        i += 1
        while data[i] < guard:
            i += 1
            if i >= h:break

        j -= 1
        while data[j] > guard:
            j -= 1
            if l >= j:break

        if j <= i: break
        data[i], data[j] = data[j], data[i]

    data[l], data[i-1] = data[i-1], data[l]
    return i


def qsort(data, l, h):
    if l >= h: return

    k = partition(data, l, h)
    print ' '.join([str(x) for x in data])
    #qsort(data, l, k-1)
    #qsort(data, k+1, h)


if __name__ == "__main__":

    test = "B A A A B A A A B A A B "

    #data = [int(x) for x in test.split()]
    data = test.split()

    qsort(data, 0, len(data)-1)

File: algorithms/quick_sort.c
/*
 * =====================================================================================
 *
 *       Filename:  quick_sort.c
 *
 *    Description:  c语言的快速排序实现
 *
 *        Version:  1.0
 *        Created:  2014年10月17日 20时58分47秒
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  杨秀隆
 *   Organization:  
 *
 * =====================================================================================
 */
#include <stdlib.h>
#include <stdio.h>

void pyprint(int *data, int len)
{
    int i;
    for (i = 1; i <= len; i++)
        printf("%d%s", data[i-1], i==len?"\n":" ");
}

void selection(int *data, int l, int h)
{

}

void swap(int *data, int i, int j)
{
    int tmp = data[i];
    data[i] = data[j];
    data[j] = tmp;
}

int partition(int *data, int l, int h)
{
    int i = l, j = h+1;

    while (1)
    {
        while (data[++i] < data[l])
            if (i >= h) break;

        while (data[--j] > data[l])
            if (l >= j) break;

        if (i >= j) break;

        swap(data, i, j);
    }

    swap(data, l, j);
    return j;

}

void quick_sort(int *data, int l, int h)
{
    selection(data, l, h);
    
    int k = partition(data, l, h);
    pyprint(data, h+1);
}

void my_qsort(int *data, int len)
{
    quick_sort(data, 0, len-1);
}

int main(int argc, char **argv)
{
    int i;
    int data[20];

    for (i = 1; i < argc; i++) 
        data[i-1] = atoi(argv[i]);

    my_qsort(data, argc-1);

}
